include/CObjectCache.h

```diff
@@ -9,58 +9,110 @@
 {
 
 namespace impl
 {
+    struct FORCE_EMPTY_BASE_OPT CMultiCache_tag {};
+
     template<template<typename...> class, template<typename...> class>
     struct is_same_templ : std::false_type {};
 
     template<template<typename...> class T>
     struct is_same_templ<T, T> : std::true_type {};
 
+    template<template<typename...> class T>
+    struct is_multi_container : std::false_type {};
+    template<>
+    struct is_multi_container<std::multimap> : std::true_type {};
+    template<>
+    struct is_multi_container<std::unordered_multimap> : std::true_type {};
+
+    template<template<typename...> class T>
+    struct is_assoc_container : std::false_type {};
+    template<>
+    struct is_assoc_container<std::map> : std::true_type {};
+    template<>
+    struct is_assoc_container<std::unordered_map> : std::true_type {};
+    template<>
+    struct is_assoc_container<std::multimap> : std::true_type {};
+    template<>
+    struct is_assoc_container<std::unordered_multimap> : std::true_type {};
+    
+    template<typename K, typename...>
+    struct PropagKeyTypeTypedef_ { using KeyType = K; };
+    template<typename ...K>
+    struct PropagKeyTypeTypedef : PropagKeyTypeTypedef_<K...> {};
+
+    template<typename T, typename ...K>
+    struct PropagTypedefs : PropagKeyTypeTypedef<K...> { using CachedType = T; };
+
     template<
         template<typename...> class ContainerT_T,
         typename T, //value type for container
         typename ...K //optionally key type for std::map/std::unordered_map
     >
     struct CObjectCacheBase
     {
+        using UnderlyingContainerType = ContainerT_T<K..., T>;
+        using IteratorType = typename UnderlyingContainerType::iterator;
+        using ConstIteratorType = typename UnderlyingContainerType::const_iterator;
+        using RevIteratorType = typename UnderlyingContainerType::reverse_iterator;
+        using ConstRevIteratorType = typename UnderlyingContainerType::const_reverse_iterator;
+
     protected:
-        using ContainerT = ContainerT_T<K..., T>;
+        using ContainerT = UnderlyingContainerType;
         ContainerT m_container;
 
-        using ValueType = typename ContainerT_T<K..., T>::value_type::second_type; // container's value_type is always instantiation of std::pair
-        using NoPtrValueType = typename std::remove_pointer<ValueType>::type;
-        using PtrToConstVal_t = const NoPtrValueType*; // ValueType is always pointer type
+        // typedefs for implementation only
+        using PairType_impl = typename ContainerT::value_type;
+        //! Always pointer type
+        using ValueType_impl = typename PairType_impl::second_type; // container's value_type is always instantiation of std::pair
+        static_assert(std::is_pointer<ValueType_impl>::value, "ValueType_impl must be pointer type!");
+        using KeyType_impl = typename PairType_impl::first_type;
+        using NoPtrValueType_impl = typename std::remove_pointer<ValueType_impl>::type;
```

Comment:
you have no use for these 2 "no pointer" types.

Reply:
I have, a line below. Readability purposes